trino/trino.go

```diff
@@ -948,56 +948,62 @@ func (qr *driverRows) fetch(allowEOF bool) error {
 	}
 	hs := make(http.Header)
 	hs.Add(trinoUserHeader, qr.stmt.user)
-	req, err := qr.stmt.conn.newRequest("GET", qr.nextURI, nil, hs)
-	if err != nil {
-		return err
-	}
-	resp, err := qr.stmt.conn.roundTrip(qr.ctx, req)
-	if err != nil {
-		if qr.ctx.Err() == context.Canceled {
-			qr.Close()
+	for qr.nextURI != "" {
+		req, err := qr.stmt.conn.newRequest("GET", qr.nextURI, nil, hs)
```

Comment:
Maybe the loop body should be extracted into another private function, it's pretty big.

Also do you think you could write a test for reading a large number of rows? I wonder if it wouldn't take too long to run.

Reply:
We tried adding the test but failed to make it run in an acceptable timeframe.

As the test is very specific to the code fix, it is not really beneficial to have it included in the test suite, but at least we have validated that the code indeed fixes the stack overflow error.

```
// This test ensures that the fetch method is not generating stack overflow errors.
// === RUN   TestFetchNoStackOverflow
// runtime: goroutine stack exceeds 1000000000-byte limit
// runtime: sp=0x14037b00390 stack=[0x14037b00000, 0x14057b00000]
// fatal error: stack overflow
func TestFetchNoStackOverflow(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping test in short mode.")
	}
	count := 0
	var buf *bytes.Buffer
	var ts *httptest.Server
	ts = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if count <= 999999 {
			if buf == nil {
				buf = new(bytes.Buffer)
				json.NewEncoder(buf).Encode(&stmtResponse{
					NextURI: ts.URL + "/v1/statement/20210817_140827_00000_arvdv/1",
				})
			}
			w.WriteHeader(http.StatusOK)
			w.Write(buf.Bytes())
			count++
			return
		}
		w.WriteHeader(http.StatusOK)
		json.NewEncoder(w).Encode(&stmtResponse{
			Error: stmtError{
				ErrorName: "TEST",
			},
		})
	}))

	db, err := sql.Open("trino", ts.URL)
	require.NoError(t, err)

	t.Cleanup(func() {
		assert.NoError(t, db.Close())
	})

	_, err = db.Query("SELECT 1")
	assert.IsTypef(t, new(ErrQueryFailed), err, "unexpected error: %w", err)
}
```

Reply:
If you add `debug.SetMaxStack(20 * 1024)` at the beginning of the test and lower the number of responses from 999999 to only 99, the test fails on master and passes in this branch and takes about 200ms, which IMO is acceptable.
Using `debug.SetMemoryLimit(100 * 1024 * 1024)` from Go 1.19 does not have any effect as the issue is about the stack size.

Reply:
Also, this is going to be handy to use as a benchmark. I did some memory profiling, and we're doing too many allocations, making the Go driver fall behind Python and Java. There's also an issue with loading the time zones when parsing timestamps. This is out of scope of this PR, but this test is valuable.

Reply:
> If you add debug.SetMaxStack(20 * 1024) at the beginning of the test and lower the number of responses from 999999 to only 99, the test fails on master and passes in this branch and takes about 200ms, which IMO is acceptable.

For some reason it locally failed on me in both branches with these numbers.  `debug.SetMaxStack(50 * 1024)` and 50 requests did fail on master but work on this branch. Just to be sure maybe retest this config on your side (to make sure we are not introducing a flaky test).